### PR TITLE
elfutils: split debuginfod into its own output

### PR DESCRIPTION
**Pull request by lheckemann**
###### Description of changes
This avoids a dependency on gcc, reducing closure size significantly for dependents (including systemd, via libbpf).

###### Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [23.11 Release Notes (or backporting 23.05 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2305-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

### CODE REVIEW

File: pkgs/development/tools/misc/elfutils/default.nix [hunk 1]
```diff
@@ -41,14 +41,16 @@ stdenv.mkDerivation rec {
 
   postPatch = ''
     patchShebangs tests/*.sh
+  '' + lib.optionalString enableDebuginfod ''
```
Comment (cole-h) [inline pkgs/development/tools/misc/elfutils/default.nix:44]: Any chance you could add a comment around here why this is desired? (Bonus points for adding it to the commit message, but IMHO a note around here is much more useful / important.)

**Review by cole-h — APPROVED** [review]

**Comment by trofi** [comment]
Bisect says d7ccc25142f62b4c25a75ae016a1ff552cc446d4 `elfutils: split debuginfod into its own output` broke aat least `gdb` and `linuxPackages_latest.systemtap` in `staging`.

Is there practical difference in closure size? `systemd` is not an optional dependency on `NixOS`. And `elfutils` is an development-style dependency. It sounds reasonable to have a bit closure around it.

**Comment by trofi** [comment]
https://github.com/NixOS/nixpkgs/pull/223240 has a bit of discussion why it should be fine to uncoinditionally enable `debuginfod`.

**Comment by lheckemann** [comment]
It makes a relatively big practical difference because it's the only thing making systemd depend on gcc. But yes, let's try to unbreak gdb :sweat_smile:

**Comment by trofi** [comment]
> It makes a relatively big practical difference because it's the only thing making systemd depend on gcc.

`debuginfod` should not depend on full `gcc` clousure. If it does I think it's a bug and it should be fixed instead of elfutils.

Can you be a bit more specific here? Which `gcc` output you are concerned about? Some `gcc` outputs are used everywhere, like `libgcc` or `libstdc++`.

```
$ nix build github:NixOS/nixpkgs/staging-next#systemd
$ nix path-info -rsSh github:NixOS/nixpkgs/staging-next#systemd |& fgrep gcc
/nix/store/1hgk46w2n5fkw9n5b2z1wlx7swfahwg9-xgcc-12.3.0-libgcc                   139.1K  139.1K
/nix/store/d4wb4vw0i85xyy7qs82l4s4mr2yigvc9-gcc-12.3.0-libgcc                    139.1K  139.1K
/nix/store/75622q2wranjp6zl81ldq5j54s788dqz-gcc-12.3.0-lib                         7.5M   38.8M
```

These are not rare dependencies.

**Comment by vcunat** [comment]
> let's try to unbreak those

By that you mean to revert this PR for now or something else?

**Comment by fabianhjr** [comment]
Trying this patch on gdb:

```
diff --git a/pkgs/development/tools/misc/gdb/default.nix b/pkgs/development/tools/misc/gdb/default.nix
index 129b55e7409..528f6cd3719 100644
--- a/pkgs/development/tools/misc/gdb/default.nix
+++ b/pkgs/development/tools/misc/gdb/default.nix
@@ -57,7 +57,7 @@ stdenv.mkDerivation rec {
   buildInputs = [ ncurses readline gmp mpfr expat libipt zlib zstd guile sourceHighlight ]
     ++ lib.optional pythonSupport python3
     ++ lib.optional doCheck dejagnu
-    ++ lib.optional enableDebuginfod (elfutils.override { enableDebuginfod = true; });
+    ++ lib.optional enableDebuginfod (elfutils.override { enableDebuginfod = true; }).debuginfod;
 
   propagatedNativeBuildInputs = [ setupDebugInfoDirs ];
 
```

**Comment by evrim** [comment]
> -    ++ lib.optional enableDebuginfod (elfutils.override { enableDebuginfod = true; });
> +    ++ lib.optional enableDebuginfod (elfutils.override { enableDebuginfod = true; }).debuginfod;

Fwiw, this doesnt solve the problem, still getting 
```                                                                                                                                    
checking for libdebuginfod >= 0.179... no                                                                                                                                                                                                      
configure: error: "--with-debuginfod was given, but libdebuginfod is missing or unusable."            
```

Tried with autoreconfHook269.

```
$ echo $buildInputs                                                                                                                                                                                           
/nix/store/4hd4jjhnmkxc745d3ca48xkjp0svb033-ncurses-6.4-dev /nix/store/4p2mz1ir0dzcnyzkzlsb7jd3jfmkba1y-readline-8.2p1-dev /nix/store/amf1l89q7b10vxdi4svamxi55yp3mvna-gmp-with-cxx-6.2.1-dev /nix/store/w4b6zkk4inxck8w904aigsh6sm1gyfi3-mpfr\
-4.2.0-dev /nix/store/v2ih2chpb47kmmdj8jga2asp3iaqgkgn-expat-2.5.0-dev /nix/store/yqd7xbwkwnj6g3bxw8vfsx42l50iqriv-libipt-2.0.5 /nix/store/w99lhkib7z3hx9lcmlz17apzcm1qpsr7-zlib-1.2.13-dev /nix/store/dzfh5wzcgs2jyk2y6l9ay6m5ds936n2b-zstd-1\
.5.5-dev /nix/store/civgvncfva22yszmq47ikx2dmy95mnf2-source-highlight-3.1.9-dev /nix/store/5a457g4zr7q11b4mz0ifc9z88kwcp8g7-python3-3.10.12 /nix/store/mk39f9yi6cpwcdlrx05ycaia0zaxk9qr-elfutils-0.189-debuginfod                              
```

**Comment by evrim** [comment]
OK, so the problem I've spotted is basically override({ enableDebugInfod = true; }) generates the following .pc:

```
$ cat /nix/store/b214w1i9xi7y16arsr1fyhypwmwk0hp7-elfutils-0.189-dev/lib/pkgconfig/libdebuginfod.pc                                                                                                                           
prefix=/nix/store/a839cj2hnqld17jcmbma4vsgr2pphaiw-elfutils-0.189                                                                                                                                                                              
exec_prefix=${prefix}                                                                                                                                                                                                                          
libdir=/nix/store/a839cj2hnqld17jcmbma4vsgr2pphaiw-elfutils-0.189/lib                                                                                                                                                                          
includedir=/nix/store/b214w1i9xi7y16arsr1fyhypwmwk0hp7-elfutils-0.189-dev/include                                                                                                                                                              
                                                                                                                                                                                                                                               
Name: debuginfod                                                                                                                                                                                                                               
Description: elfutils library to query debuginfo files from debuginfod servers                                                                                                                                                                 
Version: 0.189                                                                                                                                                                                                                                 
URL: http://elfutils.org/                                                                                                                                                                                                                      
                                                                                                                                                                                                                                               
Libs: -L${libdir} -ldebuginfod                                                                                                                                                                                                                 
Cflags: -I${includedir}                                                                                                                                                                                                                        
```

The paths the file points don't have the debuginfo libs or includes so gdb configure script fails.

**Comment by trofi** [comment]
Proposed the revert against `staging-next` as https://github.com/NixOS/nixpkgs/pull/237595

**Pull request merged**